academic_observatory_workflows/workflows/pubmed_telescope.py

```diff
@@ -0,0 +1,1590 @@
+
+
+def parse_articles(data: Dict) -> Union[List, List[Dict]]:
+    try:
+        return data["PubmedArticle"]
+    except KeyError:
+        logging.info(f"No PubmedArticle records in file")
+        return []
+
+
+def parse_deletes(data: Dict) -> Union[List, List[Dict]]:
+    try:
+        return data["DeleteCitation"]["PMID"]
+    except KeyError:
+        logging.info(f"No DeleteCitation.PMID records in file")
+        return []
+
+
+def transform_pubmed(input_path: str, upsert_path: str) -> Union[bool, str, PubmedUpdatefile]:
+    """
+    Convert a single Pubmed XML file to JSONL, pulling out any of the Pubmed the upserts and or deletes.
+    Used in parallelised transform task.
+
+    :param input_path: Path to the donwloaded xml.gz file.
+    :param upsert_path: Output file path for the upserts.
+    :return: filename of the baseline file or a PubmedUpdatefile object if it is an updatefile.
+    """
+
+    try:
+        data = load_datafile(input_path)
+    except ValidationError:
+        logging.info(f"Fields in XML are not valid against it's own DTD file - {input_path}")
+        return False
+
+    upserts = parse_articles(data)
+    logging.info(f"Pulled out {len(upserts)} upserts from file, writing to file: {upsert_path}")
+
+    # Save upserts to file.
+    save_pubmed_jsonl(upsert_path, upserts)
+
+    # Pull out keys of upserts and deletes from an updatefile. This is not required for baseline files.
+    if "upsert" in upsert_path:
+        delete_keys = parse_deletes(data)
+        logging.info(f"Pulled out {len(delete_keys)} deletes from file.")
+
+        upsert_keys = [record["MedlineCitation"]["PMID"] for record in upserts]
+
+        return PubmedUpdatefile(
+            name=os.path.basename(input_path),
+            upserts=[PMID.from_dict(record) for record in upsert_keys],
+            deletes=[PMID.from_dict(record) for record in delete_keys],
+        )
+
+    return os.path.basename(input_path)
+
+
+def save_pubmed_merged_upserts(
+    filename: str,
+    upserts: List[str],
+    input_path: str,
+    output_path: str,
+) -> str:
+    """Used in parallel by save_merged_upserts_and_deletes to write out the merged upsert records using the custom
+    Pubmed encoder.
+
+    :param upserts: A list of upsert keys to pull out and write to file.
+    :param input_path: Where the original upsert records are stored.
+    :param output_path: Destination of where to write the merged upsert records.
+    :return filename: Name of the original updatefile, for logging purposes.
+    """
+
+    full_upsert_records = {str(record["MedlineCitation"]["PMID"]): record for record in load_jsonl(input_path)}
+    upserts_to_write = [full_upsert_records[record_key] for record_key in upserts]
+    save_pubmed_jsonl(output_path, upserts_to_write)
+
+    return filename
+
+
+def merge_upserts_and_deletes(updatefiles: List[PubmedUpdatefile]) -> Tuple[Dict[str, List[str]], List[Dict]]:
```

Comment:
```suggestion
def merge_upserts_and_deletes(updatefiles: List[PubmedUpdatefile]) -> Tuple[Dict[PMID, str], Set[PMID]]:
```

Reply:
Better to keep the dataclass versions around rather than the dict versions as these are easier to work with